### DISCUSSION TITLE
Changing fallback font

### DISCUSSION THREAD

**Discussion by timoostrich** [fabricjs/fabric.js#7049] [Q&A]
I really need to make sure that the user can't use the default fallback font Times New Roman due to license problems.

How can I be sure that it will **never** appear as the fallback font?

I can override prototype:
`fabric.Text.prototype.fontFamily = "Open Sans";`

But creating a new textbox will override it.

So I can do that (multiple fonts in the fontFamily list):
`var textbox = new fabric.IText('text', {
  fontFamily: "Afont, Open Sans",
});`

That works for newly created text boxes.
But when I am loading a canvas from JSON, there won't be that fallback.

Furthermore it may happen, that Open Sans isn't loaded and the fallback will be displayed again.

What can I do to create another fallback font? Or to make the fallback just display empty boxes for example?

**Comment by asturur** [comment]
The only real issue is the font non loaded. You can't do much about it, you have to be sure it gets loaded properly.
The prototype override should work and shouldn't be overridden by anything else, or it could just be that you are doing it wrong somehow.

**Reply by timoostrich** [reply]
When I add a new text box after overriding the prototype the fallback font still is Times New Roman.

Another problem would be that my fallback font (for example Open Sans) does not contain a special character. Again there will be the Times New Roman fallback. :/

**Reply by asturur** [reply]
The TimesNewRoman fallback is os specific and Javascript has no effect there.
You have to be sure to load the font properly.

**Reply by timoostrich** [reply]
Ah, okay. I am using font preloading so that's no problem.
Thanks again for taking the time to answer my questions!

**Comment by timoostrich** [comment]
I found a way that works. But I need one last hint on how to use my solution without editing fabric.js file :-)

1. There is an [Adobe noDef](https://github.com/adobe-fonts/adobe-notdef) font which you can use as a fallback. It contains a "no def" symbol for every glyph. And it's licensed under OFL.
2. Now we need to be sure that it's always used as the fallback font on canvas.
It's possible to set multiple fonts for canvas ctx.font. Setting the noDef font as the second font works perfect but I had to add it in fabric.js directly (in the _getFontDeclaration function I added "fontFamily += ", and-regular";).

But have no idea how to overwrite that function from outside. Is that possible @asturur ? I'm no pro coder, sorry.

**Reply by asturur** [reply]
you can override getFontDeclaration is not a big deal. Is probably not changing soon nor later.

```js
fabric.Text.prototype. _getFontDeclaration = () {
  // run your code.
  // or run the original code and a small difference after.
}
```

I think you can also write a setter for the fontFaimily using Object.defineProperty, and in the set part, you append the fallback font to whatever the user choose.

**Reply by timoostrich** [reply]
Works perfect, thank you so much!  I did not know I can override the prototype like that.

In case someone else has the same problem:
I just added a new line after the fontFamily definition:
`fontFamily += ", and-regular";`

The "and-regular" font is loaded via fontFaceObserver on page load.
